crates/fuel-core/src/graphql_api/indexation.rs

```diff
@@ -0,0 +1,709 @@
+{
+    let key = message.recipient();
+    let storage = block_st_transaction.storage::<MessageBalances>();
+    let MessageBalance {
+        retryable,
+        non_retryable,
+    } = *storage.get(key)?.unwrap_or_default();
+    let current_balance = if message.has_retryable_amount() {
+        retryable
+    } else {
+        non_retryable
+    };
+
+    current_balance
+        .checked_sub(message.amount() as u128)
+        .ok_or_else(|| IndexationError::MessageBalanceWouldUnderflow {
+            owner: *message.recipient(),
+            current_amount: current_balance,
+            requested_deduction: message.amount() as u128,
+            retryable: message.has_retryable_amount(),
+        })
+        .and_then(|new_amount| {
+            let storage = block_st_transaction.storage::<MessageBalances>();
+            let new_balance = if message.has_retryable_amount() {
+                MessageBalance {
+                    retryable: new_amount,
+                    non_retryable,
+                }
+            } else {
+                MessageBalance {
+                    retryable,
+                    non_retryable: new_amount,
+                }
+            };
+            storage.insert(key, &new_balance).map_err(Into::into)
+        })
+}
+
+fn increase_coin_balance<T>(
+    block_st_transaction: &mut T,
+    coin: &Coin,
+) -> Result<(), IndexationError>
+where
+    T: OffChainDatabaseTransaction,
+{
+    let key = CoinBalancesKey::new(&coin.owner, &coin.asset_id);
+    let storage = block_st_transaction.storage::<CoinBalances>();
+    let current_amount = *storage.get(&key)?.unwrap_or_default();
+    let new_amount = current_amount.saturating_add(coin.amount as u128);
+
+    let storage = block_st_transaction.storage::<CoinBalances>();
+    Ok(storage.insert(&key, &new_amount)?)
+}
+
+fn decrease_coin_balance<T>(
+    block_st_transaction: &mut T,
+    coin: &Coin,
+) -> Result<(), IndexationError>
+where
+    T: OffChainDatabaseTransaction,
+{
+    let key = CoinBalancesKey::new(&coin.owner, &coin.asset_id);
+    let storage = block_st_transaction.storage::<CoinBalances>();
+    let current_amount = *storage.get(&key)?.unwrap_or_default();
+
+    current_amount
+        .checked_sub(coin.amount as u128)
+        .ok_or_else(|| IndexationError::CoinBalanceWouldUnderflow {
+            owner: coin.owner,
+            asset_id: coin.asset_id,
+            current_amount,
+            requested_deduction: coin.amount as u128,
+        })
+        .and_then(|new_amount| {
+            block_st_transaction
+                .storage::<CoinBalances>()
+                .insert(&key, &new_amount)
+                .map_err(Into::into)
+        })
+}
+
+pub(crate) fn process_balances_update<T>(
+    event: &Event,
+    block_st_transaction: &mut T,
+    balances_enabled: bool,
+) -> Result<(), IndexationError>
+where
+    T: OffChainDatabaseTransaction,
+{
+    if !balances_enabled {
+        return Ok(());
+    }
```

Comment:
do you need to pass balances_enabled here, or could you have this check done upstream and assume that this function is called only when `balances_enabled = true`?

Reply:
I did this to keep the call-site more clear, especially because there are more indexation coming. You can have a peek here:
https://github.com/FuelLabs/fuel-core/pull/2455/files#diff-1ac4adf5631afae925791c3897c52c2fa6282bd70901fddf552e7e1f0492c447R286-R297

Also, it makes the testing whether the flag is respected a little bit easier.

However, I can still refactor this a bit if you want, but maybe in the other, [upcoming PR](https://github.com/FuelLabs/fuel-core/pull/2455).